agent/task_context.go

```diff
@@ -191,14 +192,262 @@ func (tc *taskContext) getExecTimeout() time.Duration {
+	if err := tc.taskConfig.Validate(); err != nil {
+		return nil, err
+	}
+
+	if taskGroup == "" {
+		return &commandBlock{
+			block:       command.TaskTimeoutBlock,
+			commands:    tc.taskConfig.Project.Timeout,
+			timeoutKind: callbackTimeout,
+			getTimeout:  tc.getCallbackTimeout,
+			canFailTask: false,
+		}, nil
+	}
+
+	tg := tc.taskConfig.Project.FindTaskGroup(taskGroup)
+	if tg == nil {
+		return nil, errors.Errorf("couldn't find task group '%s' in project '%s'", taskGroup, tc.taskConfig.Project.Identifier)
+	}
+	if tg.Timeout == nil {
+		// Task group timeout defaults to the project timeout settings if not
+		// explicitly set.
+		return &commandBlock{
+			block:       command.TaskTimeoutBlock,
+			commands:    tc.taskConfig.Project.Timeout,
+			timeoutKind: callbackTimeout,
+			getTimeout:  tc.getCallbackTimeout,
+			canFailTask: false,
+		}, nil
+	}
+
+	return &commandBlock{
+		block:       command.TaskTimeoutBlock,
+		commands:    tg.Timeout,
+		timeoutKind: callbackTimeout,
+		getTimeout:  tc.getTaskGroupCallbackTimeout(tg),
+		canFailTask: false,
+	}, nil
+}
+
+// getPreTimeout returns the timeout for the pre block.
+func (tc *taskContext) getPreTimeout() time.Duration {
+	tc.RLock()
+	defer tc.RUnlock()
+
+	if tc.taskConfig != nil && tc.taskConfig.Project != nil && tc.taskConfig.Project.PreTimeoutSecs != 0 {
+		return time.Duration(tc.taskConfig.Project.PreTimeoutSecs) * time.Second
+	}
+
+	return defaultPreTimeout
+}
+
+// getPostTimeout returns the timeout for the post block.
+func (tc *taskContext) getPostTimeout() time.Duration {
+	tc.RLock()
+	defer tc.RUnlock()
+
+	if tc.taskConfig != nil && tc.taskConfig.Project != nil && tc.taskConfig.Project.PostTimeoutSecs != 0 {
+		return time.Duration(tc.taskConfig.Project.PostTimeoutSecs) * time.Second
+	}
+	return defaultPostTimeout
+}
+
+// getSetupGroupTimeout returns the timeout for the setup group block.
+func (tc *taskContext) getSetupGroupTimeout(tg *model.TaskGroup) func() time.Duration {
+	return func() time.Duration {
+		tc.RLock()
+		defer tc.RUnlock()
+
+		if tg.SetupGroupTimeoutSecs != 0 {
+			return time.Duration(tg.SetupGroupTimeoutSecs) * time.Second
+		}
+		return defaultPreTimeout
+	}
+}
+
+// getSetupTaskTimeout returns the timeout for the setup task block.
+func (tc *taskContext) getSetupTaskTimeout(tg *model.TaskGroup) func() time.Duration {
+	return func() time.Duration {
+		tc.RLock()
+		defer tc.RUnlock()
+
+		if tg.SetupTaskTimeoutSecs != 0 {
+			return time.Duration(tg.SetupTaskTimeoutSecs) * time.Second
+		}
+		return defaultPreTimeout
+	}
+}
+
+// getTeardownTaskTimeout returns the timeout for the teardown task block.
+func (tc *taskContext) getTeardownTaskTimeout(tg *model.TaskGroup) func() time.Duration {
+	return func() time.Duration {
+		tc.RLock()
+		defer tc.RUnlock()
+
+		if tg.TeardownTaskTimeoutSecs != 0 {
+			return time.Duration(tg.TeardownTaskTimeoutSecs) * time.Second
+		}
+		return defaultPostTimeout
```

Comment:
https://github.com/evergreen-ci/evergreen/pull/6918 had set the defaultPostTimeout to 2hrs. I think we might be running the risk of missing important post cleanup things by setting this to the no 30 minute defaultPostTimeout. Are there any plans to put it back to 2 hours?

Reply:
I thought about it and while it's not perfect to have invisible timeouts, 30 minutes is still somewhat more generous than the original 15 minutes, and we aren't in a rush to bump it. Letting a task run longer can increase spending and increased patch waiting time for existing projects who got used to relying on the 15 minute timeout, so I think it's better to pause on bumping it further until any projects who hit the timeout today have had a time window to set any block timeouts they want to suit their needs and/or optimize their post blocks to not hit the timeout.

From [the HC stats](https://ui.honeycomb.io/mongodb-4b/environments/production/datasets/evergreen-agent/result/i77eTtodCGG), it seems like ~90 tasks hit the default timeout per day, and they're clustered around a few projects. While it doesn't sound like a lot of tasks, a small number of tasks per day stuck running for an extra few hours can add up over time.

Reply:
They are already aware that post has a slow command and have a few tickets (such as [SERVER-79972](https://jira.mongodb.org/browse/SERVER-79972) and the related tickets) to make their post task run more efficiently.